target/linux/ramips/dts/mt7628an_tuoshi_lt22m.dts

```diff
@@ -0,0 +1,186 @@
+#include "mt7628an.dtsi"
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/mtd/partitions/uimage.h>
+
+/ {
+	compatible = "tuoshi,lt22m", "mediatek,mt7628an-soc";
+	model = "Tuoshi LT22M 4G Wireless Router";
+
+	leds {
+		compatible = "gpio-leds";
+
+		wifi {
+			label = "blue:wifi";
+			gpios = <&gpio 44 GPIO_ACTIVE_LOW>;
+		};
+
+		lan0 {
+			label = "blue:lan0";
+			gpios = <&gpio 43 GPIO_ACTIVE_LOW>;
+		};
+
+		lan1 {
+			label = "blue:lan1";
+			gpios = <&gpio 42 GPIO_ACTIVE_LOW>;
+		};
+
+		2g {
+			label = "blue:2g";
+			gpios = <&gpio 2 GPIO_ACTIVE_LOW>;
+		};
+
+		3g {
+			label = "blue:3g";
+			gpios = <&gpio 0 GPIO_ACTIVE_LOW>;
+		};
+
+		4g {
+			label = "blue:4g";
+			gpios = <&gpio 37 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	gpio-export {
+		compatible = "gpio-export";
+		#size-cells = <0>;
+
+		modem_enable {
+			gpio-export,name = "modem_enable";
+			gpio-export,output = <0>;
+			gpios = <&gpio 1 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	virtual_flash {
+		compatible = "mtd-concat";
+		devices = <&fwconcat0 &fwconcat1 &fwconcat2>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				reg = <0x0 0x0>;
+				label = "firmware";
+				compatible = "openwrt,uimage", "denx,uimage";
+				openwrt,ih-magic = <0x27151967>;
+			};
+		};
+	};
+
+};
+
+&state_default {
+	gpio {
+		groups = "i2c";
+		function = "gpio";
+	};
+};
+
+&spi0 {
+	status = "okay";
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <50000000>;
+
+	partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "u-boot";
+				reg = <0x0 0x30000>;
+				read-only;
+			};
+
+			ubootenv: partition@30000 {
+				label = "u-boot-env";
+				reg = <0x30000 0x10000>;
+			};
+
+			factory: partition@40000 {
+				compatible = "nvmem-cells";
+				label = "factory";
+				reg = <0x40000 0x10000>;
+				#address-cells = <1>;
+				#size-cells = <1>;
+				read-only;
+
+				eeprom_factory_0: eeprom@0 {
+					reg = <0x0 0x400>;
+				};
+
+				macaddr_factory_4: macaddr@4 {
+					reg = <0x4 0x6>;
+				};
+				macaddr_factory_28: macaddr@28 {
+					reg = <0x28 0x6>;
+				};
+				macaddr_factory_2e: macaddr@2e {
+					reg = <0x2e 0x6>;
+				};
```

Comment:
Thank you for the feedback. I copied most from other devices and made it work with lot of trial&error. I will go through the individual changes to understand it in detail.

macaddr_factory_4 and macaddr_factory_28 are the same. I don't know why it's there twice. macaddr_factory_2e is mac address + 2, It's probably wifi. I will check it on the stock rom.

The mac of the WWAN is not in the factory partition. Seems to be within the modem itself, the device has the correct mac without any special configuration

Reply:
WiFi is 2.4 GHz provided by the mt7628 SoC chip, right?
In such cases, it's possible that ethernet MAC and WiFi MAC are identical, so maybe offset 4 is WiFi, offset 28 is LAN, and offset 2e is WAN (not used since `eth1` is the modem).

@dangowrt said that the WiFi driver reads the MAC address from the EEPROM with no need of an explicit definition; so maybe the offset 4 one can be removed altogether.

Reply:
In the original firmware, WiFi has "mac+2" which is the mac stored in macaddr_factory_2e. In my build, WiFi mac and ethernet mac are identical.

So should I remove the definitions 28 and 2e but keep a comment about them?